ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateSplitRule.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelBuilder;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelFactories;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveAggregate;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveGroupingID;
+
+/**
+ * Rule that matches an aggregate with grouping sets and splits it into an aggregate
+ * without grouping sets (bottom) and an aggregate with grouping sets (top).
+ */
+public class HiveAggregateSplitRule extends RelOptRule {
+
+  public static final HiveAggregateSplitRule INSTANCE =
+      new HiveAggregateSplitRule(HiveAggregate.class, HiveRelFactories.HIVE_BUILDER);
+
+  private HiveAggregateSplitRule(Class<? extends Aggregate> aggregateClass,
+      RelBuilderFactory relBuilderFactory) {
+    super(
+        operandJ(aggregateClass, null, agg -> agg.getGroupType() != Group.SIMPLE, any()),
+        relBuilderFactory, null);
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+    final RelBuilder relBuilder = call.builder();
+
+    // If any aggregate is distinct, bail out
+    // If any aggregate is the grouping id, bail out
+    // If any aggregate call has a filter, bail out
+    // If any aggregate functions do not support splitting, bail out
+    final List<SqlAggFunction> topAggFunctions = new ArrayList<>();
+    for (AggregateCall aggregateCall : aggregate.getAggCallList()) {
+      if (aggregateCall.isDistinct()) {
+        return;
+      }
+      if (aggregateCall.getAggregation().equals(HiveGroupingID.INSTANCE)) {
+        return;
+      }
+      if (aggregateCall.filterArg >= 0) {
+        return;
+      }
+      SqlAggFunction aggFunction =
+          HiveRelBuilder.getRollup(aggregateCall.getAggregation());
+      if (aggFunction == null) {
+        return;
+      }
+      topAggFunctions.add(aggFunction);
+    }
+
+    final ImmutableBitSet bottomAggregateGroupSet = aggregate.getGroupSet();
+    if (aggregate.getCluster().getMetadataQuery().areColumnsUnique(aggregate.getInput(), bottomAggregateGroupSet)) {
+      // Nothing to do, probably already pushed
+      return;
+    }
+
+    final ImmutableBitSet topAggregateGroupSet = ImmutableBitSet.range(0, bottomAggregateGroupSet.cardinality());
+
+    final Map<Integer, Integer> map = new HashMap<>();
+    bottomAggregateGroupSet.forEach(k -> map.put(k, map.size()));
+    ImmutableList<ImmutableBitSet> topAggregateGroupSets = ImmutableBitSet.ORDERING.immutableSortedCopy(
+        ImmutableBitSet.permute(aggregate.groupSets, map));
+
+    final List<AggregateCall> topAggregateCalls = new ArrayList<>();
+    for (int i = 0; i < aggregate.getAggCallList().size(); i++) {
+      AggregateCall aggregateCall = aggregate.getAggCallList().get(i);
```

Comment:
Instead of looping over agg call list again I believe you can do this in the first loop itself where `topAggFuctions` is created.

Reply:
I was trying to bail out quick, but you are right that creating those objects is not such a huge overhead. I have changed it accordingly.